src/main/scala/chisel3/util/SparseVec.scala

```diff
@@ -0,0 +1,228 @@
+        val baseEncodedAddress = defaultValue match {
+          case Some(_) => 1
+          case None    => 0
+        }
+
+        // The total number of output addresses that need to be encoded.
+        val encodedSize = dataMap.size + baseEncodedAddress + {
+          zeroValue match {
+            case Some(_) => 1
+            case None    => 0
+          }
+        }
+
+        val legalBitPats = dataMap.zipWithIndex.flatMap {
+          case ((_, v), i) =>
+            v.map { vv =>
+              BitPat(vv.U(addrWidth.W)) -> BitPat(d.encoding(i + baseEncodedAddress, encodedSize))
+            }
+        } ++ {
+          // If out-of-bounds behavior is possible, then add BitPats for all
+          // out-of-bounds accesses.
+          zeroValue match {
+            case None => Seq.empty
+            case Some((key, _)) =>
+              (size until BigInt(addrWidth).pow(2).toInt).map {
+                case i =>
+                  BitPat(i.U(addrWidth.W)) -> BitPat(d.encoding(baseEncodedAddress, encodedSize))
+              }
+          }
+        }
+
+        // Generate the truth table.  Use a default value of zero.
+        val ttable = TruthTable(
+          legalBitPats,
+          BitPat.N(encodedSize)
+        )
+
+        // Generate the lookup.
+        d.lookup(decoder(addr, ttable), VecInit((defaultValue ++ dataMap.keys).toSeq))
+
+      // Generate a lookup using a sequence of when statements.
+      case SparseVec.IfElse =>
+        val result = Wire(gen)
+
+        // The default value _must_ be specified to bypass initialization checking.
+        result := 0.U
+
+        // Generate one when statement for each value.  Combine index checking
+        // into an or reduction.
+        dataMap.foreach {
+          case (k, v) =>
+            when(VecInit(v.map(addr === _.U).toSeq).reduce(_ || _)) {
+              result := k
+            }
+        }
+
+        // Return the zero value on out-of-bounds indices if these are possible.
+        zeroValue match {
+          case Some((_, v)) =>
+            when(addr >= size.U) {
+              result := v
+            }
+          case None =>
+        }
+        result
+    }
+
+    // Return a read-only value to avoid user mistakes where they try to write
+    // to a SparseVec.
+    Detail.readOnly(result)
+  }
+
+}
+
+/** Utility for creating Sparse Vectors that behave identical to [[Vec]]s.
+  *
+  * A [[SparseVec]] is like a [[Vec]] that only has some subset of its constituent elements set to values.
+  */
+object SparseVec {
+
+  /** The root type of how a [[SparseVec$]] can be accessed. */
+  sealed trait Type
+
+  /** A [[SparseVec$]] that is accessed using a [[chisel3.util.experimental.decode.decoder]]. */
+  sealed trait Decoder extends Type {
+    def encoding(index: Int, width: Int): UInt
+
+    def lookup[A <: Data](
+      index:  UInt,
+      values: VecLike[A]
+    )(
+      implicit _sourceInfo: SourceInfo
```

Comment:
```suggestion
      implicit sourceInfo: SourceInfo
```
I wouldn't use an `_` in the name here, we don't in the rest of the codebase and the name of the argument is part of the public of the function.

Reply:
Changed this over.